File: AIReaderKindle/src/Domain/AI/ChatMessage.cpp
#include "ChatMessage.hpp"

Json ChatMessage::toJson() const {
    Json json = Json::object();
    json.set("role", role);
    json.set("content", content ? Json(*content) : Json(nullptr));
    if (!toolCalls.empty()) {
        Json calls = Json::array();
        for (const auto& call : toolCalls) {
            Json function = Json::object();
            function.set("name", call.name);
            function.set("arguments", call.arguments);
            Json entry = Json::object();
            entry.set("id", call.id);
            // Services require this back on every call they made, even the
            // ones that leave it out of their own reply.
            entry.set("type", "function");
            entry.set("function", function);
            calls.push(entry);
        }
        json.set("tool_calls", calls);
    }
    if (!toolCallId.empty()) json.set("tool_call_id", toolCallId);
    return json;
}

ChatMessage ChatMessage::fromJson(const Json& json) {
    ChatMessage message;
    message.role = json.at("role").string();
    if (json.at("content").isString()) message.content = json.at("content").string();
    for (const auto& call : json.at("tool_calls").items()) {
        const Json& function = call.at("function");
        const Json& arguments = function.at("arguments");
        message.toolCalls.push_back(ToolCall{
            call.at("id").string(),
            function.at("name").string(),
            // Some services return the arguments as an object rather than a string.
            arguments.isString() ? arguments.string() : arguments.dump(),
        });
    }
    message.toolCallId = json.at("tool_call_id").string();
    return message;
}

File: AIReader/AIReader/Domain/AI/ChatMessage.swift
import Foundation

/// One message in a chat completion exchange.
struct ChatMessage: Codable, Equatable, Sendable {
    struct ToolCall: Codable, Equatable, Sendable {
        struct Function: Codable, Equatable, Sendable {
            var name: String
            /// A JSON object, encoded as a string, as the API returns it.
            var arguments: String
        }

        var id: String
        /// Services require this back on every call they made, even the ones
        /// that leave it out of their own reply.
        var type = "function"
        var function: Function
    }

    var role: String
    var content: String?
    var toolCalls: [ToolCall]?
    var toolCallID: String?

    enum CodingKeys: String, CodingKey {
        case role
        case content
        case toolCalls = "tool_calls"
        case toolCallID = "tool_call_id"
    }

    static func system(_ content: String) -> Self { Self(role: "system", content: content) }
    static func user(_ content: String) -> Self { Self(role: "user", content: content) }

    static func toolResult(_ content: String, callID: String) -> Self {
        Self(role: "tool", content: content, toolCallID: callID)
    }
}

extension ChatMessage.ToolCall {
    // Written by hand so a missing `type` in the reply falls back to the
    // default rather than failing the whole decode.
    init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        type = try container.decodeIfPresent(String.self, forKey: .type) ?? "function"
        function = try container.decode(Function.self, forKey: .function)
    }
}
